src/platform_impl/x11/mod.rs

```diff
@@ -94,333 +101,396 @@ impl Drop for GlobalHotKeyManager {
-                                                    id: *id,
-                                                    state: crate::HotKeyState::Pressed,
-                                                });
-                                                *pressed = true;
-                                            }
-                                        }
-                                    }
-                                    xlib::KeyRelease => {
-                                        for (id, _, pressed) in entry {
-                                            if *pressed {
-                                                GlobalHotKeyEvent::send(GlobalHotKeyEvent {
-                                                    id: *id,
-                                                    state: crate::HotKeyState::Released,
-                                                });
-                                                *pressed = false;
-                                            }
-                                        }
-                                    }
-                                    _ => {}
-                                }
+fn events_processor(thread_rx: Receiver<ThreadMessage>) -> Result<(), String> {
+    //                           key           id,  mods,    pressed
+    let mut hotkeys = BTreeMap::<Keycode, Vec<(u32, ModMask, bool)>>::new();
+
+    let (conn, screen) = RustConnection::connect(None)
+        .map_err(|err| format!("Unable to open x11 connection, maybe you are not running under X11? Other window systems on Linux are not supported by `global-hotkey` crate: {}", err))?;
+
+    xkb::ConnectionExt::xkb_use_extension(&conn, 1, 0)
+        .map_err(|err| {
+            format!(
+                "Unable to send xkb_use_extension request to x11 server: {}",
+                err
+            )
+        })?
+        .reply()
+        .map_err(|err| {
+            format!(
+                "xkb_use_extension request to x11 server has failed: {}",
+                err
+            )
+        })?;
+
+    xkb::ConnectionExt::xkb_per_client_flags(
+        &conn,
+        xkb::ID::USE_CORE_KBD.into(),
+        xkb::PerClientFlag::DETECTABLE_AUTO_REPEAT,
+        xkb::PerClientFlag::DETECTABLE_AUTO_REPEAT,
+        Default::default(),
+        Default::default(),
+        Default::default(),
+    )
+    .map_err(|err| {
+        format!(
+            "Unable to send xkb_per_client_flags request to x11 server: {}",
+            err
+        )
+    })?
+    .reply()
+    .map_err(|err| {
+        format!(
+            "xkb_per_client_flags request to x11 server has failed: {}",
+            err
+        )
+    })?;
+
+    let root = conn.setup().roots[screen].root;
+
+    let full_mask = KeyButMask::CONTROL | KeyButMask::SHIFT | KeyButMask::MOD4 | KeyButMask::MOD1;
+
+    loop {
+        while let Ok(Some(event)) = conn.poll_for_event() {
```

Comment:
maybe we should change this to use `wait_for_event` and avoid polling and sleeping

Reply:
I thought about doing this, but the problem is that we wait on 2 different blocking API changing that is non trivial so I left it as is

Reply:
I can try to tackle is in a separate pr but for now I want to merge this asap

Reply:
thank you